README
NAME
    Error::Return - return(), skipping a scope

VERSION
    version 1.100850

SYNOPSIS
        use Error ':try';
        use Error::Return;

        sub doit {
            print " in doit, before try\n";
            try {
                print "  in try: start\n";
                RETURN 456;
                print "  in try: end\n";
            } catch Error with {
                my $E = shift;
                print "  caught error [$E]\n";
            };
            print " in doit, after try\n";
        }

        print "before doit\n";
        my $x = doit();
        print "doit() returned [$x]\n";
        print "after doit\n";

        # prints:
        #
        # before doit
        #  in doit, before try
        #   in try: start
        # doit() returned [456]
        # after doit

DESCRIPTION
    The Error module provides "try()", which takes a coderef using the "&"
    prototype so it looks more like a normal Perl block or like "map()" or
    "grep()". But the "block" is still just an anonymous subroutine, so
    using "return" within the sub won't do what you think it will do. For
    example:

        sub doit {
            print " in doit, before try\n";
            try {
                print "  in try: start\n";
                return 456;
                print "  in try: end\n";
            } catch Error with {
                my $E = shift;
                print "  caught error [$E]\n";
            };
            print " in doit, after try\n";
        }

        print "before doit\n";
        my $x = doit();
        print "doit() returned [$x]\n";
        print "after doit\n";

    The "return" in the "try" block (we call it a block, but it really
    isn't) looks like it should return from "doit()", but it doesn't - it
    just returns from the anonymous sub that was passed to "try()".
    Therefore, this program prints the following:

        before doit
         in doit, before try
          in try: start
         in doit, after try
        doit() returned [1]
        after doit

    So "in doit, after try" is still reached, and "doit()" returns 1 because
    of its last "print" statement.

    While that is the correct behaviour, it is unintuitive. This module
    provides a more powerful way of returning.

METHODS
  RETURN
    Like "return" except that it doesn't just return to its upper scope but
    smashes right through it to the next-higher scope. Actually, it skips
    two scopes, because it has to return from the "try()" subroutine as
    well. It does take care of the cleanup that "try()" would normally
    perform.

    See the synopsis as an example - this way, the "try" block will "do what
    you mean".

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Error-Return>.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see <http://search.cpan.org/dist/Error-Return/>.

    The development version lives at
    <http://github.com/hanekomu/Error-Return/>. Instead of sending patches,
    please fork this project using the standard git and github
    infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2009 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Error-Return

1.100850  2010-03-26 23:59:47 Europe/Vienna
    - converted the distribution to Dist::Zilla-style

0.02 Sun 2009.10.25 09:19:46 CET (Marcel Gruenauer <marcel@cpan.org>)
    - required perl version is 5.008; had lots of failure reports for 5.6.2.

0.01 Fri Feb 13 15:58:07 CET 2009 (Marcel Gruenauer <marcel@cpan.org>)
    - original version


lib/Error/Return.pm
use 5.008;
use strict;
use warnings;

package Error::Return;
our $VERSION = '1.100850';
# ABSTRACT: return(), skipping a scope
use Scope::Upper qw(unwind want_at :words);
use Exporter qw(import);
our @EXPORT = qw(RETURN);

sub RETURN {
    my $context = SUB UP SUB UP SUB UP SUB;

    # Do the cleanup that try() would normally do if the try-block had ended
    # without an exception>
    shift @Error::STACK;
    unwind + (want_at($context) ? @_ : $_[0]) => $context;
}
1;


__END__
=pod

=head1 NAME

Error::Return - return(), skipping a scope

=head1 VERSION

version 1.100850

=head1 SYNOPSIS

    use Error ':try';
    use Error::Return;

    sub doit {
        print " in doit, before try\n";
        try {
            print "  in try: start\n";
            RETURN 456;
            print "  in try: end\n";
        } catch Error with {
            my $E = shift;
            print "  caught error [$E]\n";
        };
        print " in doit, after try\n";
    }

    print "before doit\n";
    my $x = doit();
    print "doit() returned [$x]\n";
    print "after doit\n";

    # prints:
    #
    # before doit
    #  in doit, before try
    #   in try: start
    # doit() returned [456]
    # after doit

=head1 DESCRIPTION

The L<Error> module provides C<try()>, which takes a coderef using the C<&>
prototype so it looks more like a normal Perl block or like C<map()> or
C<grep()>. But the "block" is still just an anonymous subroutine, so using
C<return> within the sub won't do what you think it will do. For example:

    sub doit {
        print " in doit, before try\n";
        try {
            print "  in try: start\n";
            return 456;
            print "  in try: end\n";
        } catch Error with {
            my $E = shift;
            print "  caught error [$E]\n";
        };
        print " in doit, after try\n";
    }

    print "before doit\n";
    my $x = doit();
    print "doit() returned [$x]\n";
    print "after doit\n";

The C<return> in the C<try> block (we call it a block, but it really isn't)
looks like it should return from C<doit()>, but it doesn't - it just returns
from the anonymous sub that was passed to C<try()>. Therefore, this program
prints the following:

    before doit
     in doit, before try
      in try: start
     in doit, after try
    doit() returned [1]
    after doit

So C<in doit, after try> is still reached, and C<doit()> returns C<1> because
of its last C<print> statement.

While that is the correct behaviour, it is unintuitive. This module provides a
more powerful way of returning.

=head1 METHODS

=head2 RETURN

Like C<return> except that it doesn't just return to its upper scope but
smashes right through it to the next-higher scope. Actually, it skips two
scopes, because it has to return from the C<try()> subroutine as well. It does
take care of the cleanup that C<try()> would normally perform.

See the synopsis as an example - this way, the C<try> block will "do what you
mean".

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Error-Return>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Error-Return/>.

The development version lives at
L<http://github.com/hanekomu/Error-Return/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2009 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker 6.11;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'Error::Return',
                       'DISTNAME' => 'Error-Return',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.11'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
                       'BUILD_REQUIRES' => {},
                       'ABSTRACT' => 'return(), skipping a scope',
                       'EXE_FILES' => [],
                       'VERSION' => '1.100850',
                       'PREREQ_PM' => {
                                        'Scalar::Util' => '0',
                                        'Scope::Upper' => '0',
                                        'English' => '0',
                                        'Exporter' => '0',
                                        'Carp' => '0',
                                        'Test::More' => '0.88',
                                        'Error' => '0',
                                        'File::Find' => '0',
                                        'File::Temp' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{LICENSE}
  unless eval { ExtUtils::MakeMaker->VERSION(6.31) };

WriteMakefile(%WriteMakefileArgs);



